{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='home'></a>\n",
    "### purpose\n",
    "\n",
    "- create R scripts for training and fitting gradient forest models\n",
    "- test them on a very small snps file\n",
    "\n",
    "### outline\n",
    "\n",
    "[1. Training script](#train)\n",
    "\n",
    "[2. Fitting script](#fit)\n",
    "\n",
    "[3. test training script](#test)\n",
    "\n",
    "[4. test fitting script](#testfit)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "56 56\n",
      "##################################################################\n",
      "Today:\tJune 09, 2023 - 12:38:02\n",
      "python version: 3.8.5\n",
      "conda env: newpy385\n",
      "\n",
      "Current commit of pythonimports:\n",
      "commit 03d76f7a992130f4b94ac34a09ad439e918d3044  \n",
      "Author: Brandon Lind <lind.brandon.m@gmail.com>  \n",
      "Date:   Fri Jun 9 09:42:21 2023 -0400\n",
      "##################################################################\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "balance_queue       NA\n",
       "cartopy             0.18.0\n",
       "ipyparallel         6.3.0\n",
       "matplotlib          3.3.1\n",
       "myclasses           NA\n",
       "myfigs              NA\n",
       "mymaps              NA\n",
       "myslurm             NA\n",
       "numpy               1.22.4\n",
       "pandas              1.5.3\n",
       "pythonimports       NA\n",
       "session_info        1.0.0\n",
       "tqdm                4.55.1\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "PIL                 9.4.0\n",
       "attr                22.2.0\n",
       "backcall            0.2.0\n",
       "certifi             2022.12.07\n",
       "cffi                1.14.5\n",
       "cycler              0.10.0\n",
       "cython_runtime      NA\n",
       "dateutil            2.8.2\n",
       "decorator           5.0.9\n",
       "defusedxml          0.7.1\n",
       "fiona               1.8.20\n",
       "ipykernel           5.4.2\n",
       "ipython_genutils    0.2.0\n",
       "ipywidgets          7.6.3\n",
       "jedi                0.17.0\n",
       "kiwisolver          1.4.4\n",
       "matplotlib_venn     0.11.6\n",
       "mpl_toolkits        NA\n",
       "parso               0.7.1\n",
       "pexpect             4.8.0\n",
       "pickleshare         0.7.5\n",
       "pkg_resources       NA\n",
       "prompt_toolkit      3.0.17\n",
       "ptyprocess          0.7.0\n",
       "pygments            2.9.0\n",
       "pyparsing           3.0.9\n",
       "pytz                2022.7.1\n",
       "scipy               1.6.2\n",
       "seaborn             0.11.0\n",
       "setuptools          67.2.0\n",
       "shapefile           2.1.3\n",
       "shapely             1.7.1\n",
       "six                 1.16.0\n",
       "statsmodels         0.12.2\n",
       "storemagic          NA\n",
       "tornado             6.1\n",
       "traitlets           5.0.5\n",
       "wcwidth             0.2.5\n",
       "zmq                 22.1.0\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             7.19.0\n",
       "jupyter_client      6.1.12\n",
       "jupyter_core        4.7.1\n",
       "notebook            6.4.0\n",
       "-----\n",
       "Python 3.8.5 (default, Sep  4 2020, 07:30:14) [GCC 7.3.0]\n",
       "Linux-3.10.0-1160.90.1.el7.x86_64-x86_64-with-glibc2.10\n",
       "-----\n",
       "Session information updated at 2023-06-09 12:38\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from pythonimports import *\n",
    "\n",
    "DIR = '/data/projects/pool_seq/phenotypic_data/offset_misc_files'\n",
    "\n",
    "lview,dview = get_client()\n",
    "\n",
    "latest_commit()\n",
    "sinfo(html=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='train'></a>\n",
    "# Training script\n",
    "\n",
    "[top](#home)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "training_script = '''\n",
    "#--------------------------------------------------------------------------------------------------------#\n",
    "# Given a set of populations, allele freqs, and environmental data, train gradient forests.\n",
    "#\n",
    "# Usage\n",
    "# -----\n",
    "# Rscript gradient_training.R snpfile envfile range_file basename save_dir imports_path\n",
    "#\n",
    "# Notes\n",
    "# -----\n",
    "# - snpfile and envfile must already be subset for training populations\n",
    "#\n",
    "# Parameters\n",
    "# ----------\n",
    "# snpfile - SNPs that you want to use to train gradient forests (row.names=pop_ID, cols=loci)\n",
    "# envfile - lat/long, environmental data for each pop\n",
    "# range_file - tab delimited .txt file path with columns for `lat`, `lon`, and each env used in training\n",
    "#    - this will be used to interpolate model on pops' lat/long to a larger area\n",
    "#      - specifically this should the range-wide environmental data that is from same \n",
    "#        timeframe as in `envfile` (eg 1961-1990)\n",
    "#    - generally, these are the envdata extracted from ClimateNA GIS data, and then clipped\n",
    "#        to the range map\n",
    "#    - these can be current, future, or uniform (eg for common garden) climate\n",
    "#    - see eg yeaman03:/notebooks/007_JP_gea/07_gradient_forest/01_prep_spatial_envdata.ipynb\n",
    "# basename - the basename prefix wanted for saving files in `save_dir`\n",
    "# save_dir - the directory where files should be saved\n",
    "# imports_path - path to directory with my homebrew R functions\n",
    "#--------------------------------------------------------------------------------------------------------#\n",
    "\n",
    "library(data.table)\n",
    "library(hash)\n",
    "library(gradientForest)\n",
    "library(raster)\n",
    "library(rgeos)\n",
    "\n",
    "options(datatable.fread.datatable=FALSE)\n",
    "\n",
    "args = commandArgs(trailingOnly=TRUE)\n",
    "snpfile <- args[1]\n",
    "envfile <- args[2]\n",
    "range_file <- args[3]\n",
    "basename <- args[4]\n",
    "save_dir <- args[5]\n",
    "imports_path <- args[6]\n",
    "\n",
    "source(paste0(imports_path, '/imports.R'))  # https://github.com/brandonlind/r_imports\n",
    "\n",
    "print(format(Sys.time(), format=\"%B %d %Y %T\"))\n",
    "\n",
    "prep_snps <- function(snpfile, exclude=c(), sep='\\\\t'){\n",
    "    #----------------------------------------------------------------#\n",
    "    # READ IN FREQUENCYxPOP TABLE, REMOVE POPS IN exclude\n",
    "    #\n",
    "    # Notes\n",
    "    # -----\n",
    "    # - assumes snpfile has colname 'index' which has pop names\n",
    "    # - the rest of the colnames are names for each locus in dataset\n",
    "    #----------------------------------------------------------------#\n",
    "    # load data\n",
    "    snps <- fread(snpfile, sep=sep)\n",
    "    # set row names\n",
    "    row.names(snps) <- snps[ ,'index']\n",
    "    # remove column so all cols are loci\n",
    "    snps <- subset(snps, select = -c(index))\n",
    "    # get popnames, exclude irrelevant columns\n",
    "    pops <- rownames(snps)[!rownames(snps) %in% c('CHROM', 'POS')]\n",
    "    if (len(exclude) > 0){\n",
    "        # exclude any other pops\n",
    "        pops <- pops[!pops %in% exclude]\n",
    "    }\n",
    "    # reframe datatable\n",
    "    snps <- snps[pops, ]\n",
    "    return(snps)\n",
    "}\n",
    "\n",
    "prep_envdata <- function(envfile, pops){\n",
    "    #-------------------------------------------------------------------------#\n",
    "    # COLOR GEOGRAPHIC GROUPS AND REORDER POPS IN ENVIRONMENT TABLE \n",
    "    #\n",
    "    # Notes\n",
    "    # -----\n",
    "    # - places `envs` (names of environments + elevation) in global namespace\n",
    "    #-------------------------------------------------------------------------#\n",
    "    \n",
    "    # read in the environmental data (previously centered/standardized)\n",
    "    envdata <- read.csv(envfile, sep='\\\\t', row.names=1)\n",
    "\n",
    "    envs <- c('AHM', 'CMD', 'DD5', 'DD_0', 'EMT', 'EXT', 'Eref', 'FFP', 'MAP', 'MAT', 'MCMT', 'MSP', 'MWMT',\n",
    "              'NFFD', 'PAS', 'SHM', 'TD', 'bFFP', 'eFFP')\n",
    "\n",
    "    stopifnot(all(envs %in% colnames(envdata)))\n",
    "    cat(sprintf('There are %s environmental attributes in the envdata table.', len(envs)))\n",
    "    \n",
    "    # order pops same as snp table\n",
    "    envdata <- envdata[pops, ]\n",
    "    \n",
    "    # add to global namespace so I don't have to return a stupid list, stupid R\n",
    "    envs <<- envs\n",
    "    \n",
    "    return(envdata[, envs])\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### SET UP NAMESPACE\n",
    "\n",
    "# get allele freqs for each pop\n",
    "cat(sprintf('\\nPrepping snp data ...\\n'))\n",
    "cand_freqs <- prep_snps(snpfile) #, exclude=test_pops)\n",
    "\n",
    "# get population names\n",
    "pops <- rownames(cand_freqs)\n",
    "cat(sprintf('\\nPops ...\\n'))\n",
    "cat(sprintf(pops))\n",
    "\n",
    "# prep training data (climate data at location of populations)\n",
    "cat(sprintf('\\n\\nPrepping envdata\\n ...'))\n",
    "envdata <- prep_envdata(envfile, pops)\n",
    "range_data <- read.csv(range_file, sep='\\t')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### TRAIN GRADIENT FORESTS\n",
    "\n",
    "cat(sprintf('\\n\\nTraining gradient forests ...\\n'))\n",
    "\n",
    "# double check that pops are in same order\n",
    "print(rownames(envdata))\n",
    "print(rownames(cand_freqs))\n",
    "stopifnot(all(rownames(envdata)==rownames(cand_freqs)))\n",
    "\n",
    "# NOTE: envdata below must have been filtered for env-only columns\n",
    "# calc of maxLevel is different but not effectively from Ellis 2012 Supp 1 https://doi.org/10.6084/m9.figshare.c.3304350.v1\n",
    "# recommended by Stephen Keller\n",
    "maxLevel <- floor(log2(0.368*nrow(envdata)/2))\n",
    "ptm <- proc.time()\n",
    "gfOut <- gradientForest(cbind(envdata, cand_freqs),\n",
    "                        predictor.vars=colnames(envdata[ ,envs]),\n",
    "                        response.vars=colnames(cand_freqs),\n",
    "                        ntree=500,\n",
    "                        transform=NULL,\n",
    "                        nbin=201,  # this isn't doing anything since I didn't use `compact`\n",
    "                        maxLevel=maxLevel,\n",
    "                        corr.threshold=0.5)\n",
    "print(proc.time() - ptm)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "### INTEROPOLATE MODEL TO RANGE_WIDE DATA FOR THE SAME TIME PERIOD\n",
    "cat(sprintf('\\n\\nInterpolating gradient forests model ...\\n'))\n",
    "ptm <- proc.time()\n",
    "predOut.gf <- predict(gfOut, range_data[ ,envs])\n",
    "print(proc.time() - ptm)\n",
    "\n",
    "\n",
    "\n",
    "### SAVE\n",
    "cat(sprintf('\\n\\nSaving files ...\\n'))\n",
    "gfOut_trainingfile <- paste0(\n",
    "    save_dir,\n",
    "    sprintf('/%s_gradient_forest_training.RDS', basename)\n",
    ")\n",
    "saveRDS(gfOut, gfOut_trainingfile)\n",
    "\n",
    "predfile <- paste0(\n",
    "    save_dir,\n",
    "    sprintf('/%s_gradient_forest_predOut.RDS', basename)\n",
    ")\n",
    "saveRDS(predOut.gf, predfile)\n",
    "\n",
    "print(gfOut_trainingfile)\n",
    "print(predfile)\n",
    "\n",
    "\n",
    "### END\n",
    "cat(sprintf('\\nDONE!\\n'))\n",
    "print(format(Sys.time(), format=\"%B %d %Y %T\"))\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5d5099302b049d799dac88cc88e5974e  /data/projects/pool_seq/phenotypic_data/offset_misc_files/gradient_training.R\r\n"
     ]
    }
   ],
   "source": [
    "script_file = op.join(DIR, 'gradient_training.R')\n",
    "\n",
    "with open(script_file, 'w') as o:\n",
    "    o.write(training_script)\n",
    "\n",
    "!md5sum $script_file\n",
    "# 8c1f41175f0d137fe1983ab82c3bacea - before revision1 md5sum"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='fit'></a>\n",
    "# 2. fitting script \n",
    "\n",
    "[top](#home)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "fitting_script = '''\n",
    "#----------------------------------------------------------------------------------------------------------#\n",
    "# Given a trained gradient forest, fit model to input climate data, `range_data`; save offset netCDF file.\n",
    "#\n",
    "# Usage\n",
    "# -----\n",
    "# Rscript gradient_fitting_script.R gfOut_trainingfile range_file predfile maskfile basename save_dir\n",
    "#\n",
    "# Parameters\n",
    "# ----------\n",
    "# gfOut_trainingfile - file path to trained gradient forest object, saved as .RDS\n",
    "# range_file - tab delimited .txt file path with columns for `lat`, `lon`, and each env used in training\n",
    "#    - generally, these are the envdata extracted from ClimateNA GIS data, and then clipped\n",
    "#        to the range map\n",
    "#    - these can be current, future, or uniform (eg for common garden) climate\n",
    "#    - see eg yeaman03:/notebooks/007_JP_gea/07_gradient_forest/01_prep_spatial_envdata.ipynb\n",
    "# predfile - path to RDS of object saved from interpolating gradient forest model across landscape\n",
    "#    for the same time period as the data that went into training (created in gradient_training.R)\n",
    "# maskfile - path to RDS object of a blank range map (same map used to clip range_file data)\n",
    "#    - this was created by setting all values to 0, and then saving RDS\n",
    "#    - see https://www.mountainmanmaier.com/software/pop_genom/\n",
    "# basename - the basename prefix wanted for saving files in `save_dir`\n",
    "# save_dir - the directory where files should be saved\n",
    "#----------------------------------------------------------------------------------------------------------#\n",
    "\n",
    "library(gradientForest)\n",
    "library(raster)\n",
    "\n",
    "args = commandArgs(trailingOnly=TRUE)\n",
    "gfOut_trainingfile <- args[1]\n",
    "range_file <- args[2]\n",
    "predfile <- args[3]\n",
    "maskfile <- args[4]\n",
    "basename <- args[5]\n",
    "save_dir <- args[6]\n",
    "\n",
    "calc_euclid <- function(proj, curr){\n",
    "    #-------------------------------------------------------------------#\n",
    "    # CALCULATE EUCLIDEAN DISTANCE BETWEEN CURRENT AND PROJECTED OFFSET\n",
    "    #-------------------------------------------------------------------#\n",
    "    \n",
    "    df <- data.frame(matrix(ncol=ncol(proj), nrow=nrow(proj)))\n",
    "    stopifnot(all(colnames(proj) == colnames(curr)))\n",
    "    for (i in 1:ncol(proj)){\n",
    "        # square each difference between the column data\n",
    "        df[,i] <- (proj[,i] - curr[,i])**2\n",
    "    }\n",
    "    # sum by rows then take the square root of each\n",
    "    return (sqrt(apply(df, 1, sum)))\n",
    "}\n",
    "\n",
    "get_offset <- function(outfile){\n",
    "    #--------------------------#\n",
    "    # CALCULATE GENETIC OFFSET\n",
    "    #--------------------------#\n",
    "    \n",
    "    # predict offset for common garden\n",
    "    ptm <- proc.time()\n",
    "    projOut <- predict(gfOut, range_data[, envs])\n",
    "    print(proc.time() - ptm)\n",
    "\n",
    "    # calc offset\n",
    "    offset <- calc_euclid(projOut, predOut)\n",
    "\n",
    "    # put offset into GIS layer\n",
    "    m2 <- readRDS(maskfile)\n",
    "    m2[range_cells] <- offset\n",
    "\n",
    "    # save netCDF file\n",
    "    writeRaster(m2,\n",
    "                outfile,\n",
    "                format='CDF',\n",
    "                overwrite=TRUE)\n",
    "    cat(sprintf('\\nSaved netCDF file to:\\n'))\n",
    "    cat(sprintf(sprintf('\\t%s', outfile)))\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "# SET UP NAMESPACE\n",
    "envs <- c('Elevation', 'AHM', 'CMD', 'DD5', 'DD_0', 'EMT', 'EXT', 'Eref', 'FFP', 'MAP', 'MAT', 'MCMT', 'MSP', 'MWMT',\n",
    "              'NFFD', 'PAS', 'SHM', 'TD', 'bFFP', 'eFFP')\n",
    "stopifnot(length(envs) == 20)\n",
    "gfOut <- readRDS(gfOut_trainingfile)\n",
    "range_data <- read.csv(range_file, sep='\\\\t')\n",
    "predOut <- readRDS(predfile)\n",
    "\n",
    "# load mask shapefile, use to get cell IDs for each of the lat/long in range_data\n",
    "mask <- readRDS(maskfile)\n",
    "range_cells <- cellFromXY(mask, range_data[ ,c('lon','lat')])\n",
    "\n",
    "\n",
    "\n",
    "# FIT GRADIENT FOREST MODEL TO INPUT ENVDATA\n",
    "cat(sprintf('\\nFitting gradient forest model to input envdata ...\\n'))\n",
    "ptm <- proc.time()\n",
    "projOut <- predict(gfOut, range_data[ , envs])\n",
    "print(proc.time() - ptm)\n",
    "\n",
    "\n",
    "\n",
    "# CALCULATE OFFSET\n",
    "cat(sprintf('\\nCalculating offset ...\\n'))\n",
    "netCDF_file <- paste0(\n",
    "    save_dir,\n",
    "    sprintf('/%s_offset.nc', basename)\n",
    ")\n",
    "get_offset(netCDF_file)\n",
    "\n",
    "\n",
    "# SAVE\n",
    "outfile <- paste0(\n",
    "    save_dir,\n",
    "    sprintf('/%s_gradient_forest_fitted.RDS', basename)\n",
    ")\n",
    "saveRDS(projOut, outfile)\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dc3f6d207850ad6730b539075f3a77be  /data/projects/pool_seq/phenotypic_data/offset_misc_files/gradient_fitting_script.R\r\n"
     ]
    }
   ],
   "source": [
    "fitting_file = op.join(DIR, 'gradient_fitting_script.R')\n",
    "\n",
    "with open(fitting_file, 'w') as o:\n",
    "    o.write(fitting_script)\n",
    "\n",
    "!md5sum $fitting_file\n",
    "# dc3f6d207850ad6730b539075f3a77be"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='test'></a>\n",
    "# 3. test training script\n",
    "\n",
    "[top](#home)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hash-2.2.6.1 provided by Decision Patterns\n",
      "\n",
      "\n",
      "Attaching package: ‘hash’\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    copy\n",
      "\n",
      "Loading required package: extendedForest\n",
      "extendedForest 1.6.1\n",
      "Type rfNews() to see new features/changes/bug fixes.\n",
      "Loading required package: sp\n",
      "\n",
      "Attaching package: ‘raster’\n",
      "\n",
      "The following objects are masked from ‘package:hash’:\n",
      "\n",
      "    values, values<-\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "rgeos version: 0.5-5, (SVN revision 640)\n",
      " GEOS runtime version: 3.8.0-CAPI-1.13.1 \n",
      " Linking to sp version: 1.4-5 \n",
      " Polygon checking: TRUE \n",
      "\n",
      "Warning message:\n",
      "In fun(libname, pkgname) :\n",
      "  rgeos: versions of GEOS runtime 3.8.0-CAPI-1.13.1\n",
      "and GEOS at installation 3.4.2-CAPI-1.8.2differ\n",
      "[1] \"July 02 2021 10:27:58\"\n",
      "\n",
      "Prepping snp data ...\n",
      "\n",
      "Pops ...\n",
      "DF_p3 DF_p4 DF_p5 DF_p7 DF_p8 DF_p9 DF_p18 DF_p19 DF_p20 DF_p33 DF_p34 DF_p35 DF_p36 DF_p37 DF_p38 DF_p39 DF_p40 DF_p41 DF_p42 DF_p43 DF_p44 DF_p45 DF_p46 DF_p47 DF_p48 DF_p49 DF_p54 DF_p55 DF_p56 DF_p57 DF_p58 DF_p59 DF_p60 DF_p61 DF_p62\n",
      "\n",
      "Prepping envdata\n",
      " ...There are 20 environmental attributes in the envdata table.\n",
      "\n",
      "Training gradient forests ...\n",
      " [1] \"DF_p3\"  \"DF_p4\"  \"DF_p5\"  \"DF_p7\"  \"DF_p8\"  \"DF_p9\"  \"DF_p18\" \"DF_p19\"\n",
      " [9] \"DF_p20\" \"DF_p33\" \"DF_p34\" \"DF_p35\" \"DF_p36\" \"DF_p37\" \"DF_p38\" \"DF_p39\"\n",
      "[17] \"DF_p40\" \"DF_p41\" \"DF_p42\" \"DF_p43\" \"DF_p44\" \"DF_p45\" \"DF_p46\" \"DF_p47\"\n",
      "[25] \"DF_p48\" \"DF_p49\" \"DF_p54\" \"DF_p55\" \"DF_p56\" \"DF_p57\" \"DF_p58\" \"DF_p59\"\n",
      "[33] \"DF_p60\" \"DF_p61\" \"DF_p62\"\n",
      " [1] \"DF_p3\"  \"DF_p4\"  \"DF_p5\"  \"DF_p7\"  \"DF_p8\"  \"DF_p9\"  \"DF_p18\" \"DF_p19\"\n",
      " [9] \"DF_p20\" \"DF_p33\" \"DF_p34\" \"DF_p35\" \"DF_p36\" \"DF_p37\" \"DF_p38\" \"DF_p39\"\n",
      "[17] \"DF_p40\" \"DF_p41\" \"DF_p42\" \"DF_p43\" \"DF_p44\" \"DF_p45\" \"DF_p46\" \"DF_p47\"\n",
      "[25] \"DF_p48\" \"DF_p49\" \"DF_p54\" \"DF_p55\" \"DF_p56\" \"DF_p57\" \"DF_p58\" \"DF_p59\"\n",
      "[33] \"DF_p60\" \"DF_p61\" \"DF_p62\"\n",
      "                   user                  system                 elapsed \n",
      "6.133999999999999452882 0.039000000000000006883 6.186999999999999388933 \n",
      "\n",
      "\n",
      "Interpolating gradient forests model ...\n",
      "                   user                  system                 elapsed \n",
      "0.231999999999999317879 0.032000000000000000666 0.263999999999999346301 \n",
      "\n",
      "\n",
      "Saving files ...\n",
      "[1] \"/lu213/brandon.lind/data/testdir/test_gradient_forest_training.RDS\"\n",
      "[1] \"/lu213/brandon.lind/data/testdir/test_gradient_forest_predOut.RDS\"\n",
      "\n",
      "DONE!\n",
      "[1] \"July 02 2021 10:28:07\"\n"
     ]
    }
   ],
   "source": [
    "snpfile = '/lu213/brandon.lind/data/testdir/snps_delete.txt'\n",
    "# envfile = '/data/projects/pool_seq/environemental_data/change_p6/df_interior-naturalpops_raw_env-19variables_change-p6.txt'\n",
    "envfile = '/data/projects/pool_seq/phenotypic_data/offset_misc_files/training/training_files/fdi-baypass-full_envdata.txt'\n",
    "range_file = '/data/projects/pool_seq/environemental_data/new_netCDF_files/NA_NORM_1961-1990/NA_NORM_1961-1990_all-envs_WGS84_clipped_interior.txt'\n",
    "basename = 'test'\n",
    "save_dir = '/lu213/brandon.lind/data/testdir'\n",
    "imports_path = '/lu213/brandon.lind/mypy/r_imports'\n",
    "Rscript = '/lu213/brandon.lind/data/anaconda3/envs/r35/bin/Rscript'  # r 3.5.1 env seen elsewhere in my notebooks\n",
    "\n",
    "!$Rscript $script_file $snpfile $envfile $range_file $basename $save_dir $imports_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='testfit'></a>\n",
    "\n",
    "# 4. test fitting script\n",
    "\n",
    "[top](#home)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loading required package: extendedForest\n",
      "extendedForest 1.6.1\n",
      "Type rfNews() to see new features/changes/bug fixes.\n",
      "Loading required package: sp\n",
      "\n",
      "Fitting gradient forest model to input envdata ...\n",
      "   user  system elapsed \n",
      "  0.228   0.036   0.264 \n",
      "\n",
      "Calculating offset ...\n",
      "   user  system elapsed \n",
      "  0.290   0.023   0.314 \n",
      "Loading required namespace: ncdf4\n",
      "\n",
      "Saved netCDF file to:\n",
      "\t/lu213/brandon.lind/data/testdir/something_offset.nc"
     ]
    }
   ],
   "source": [
    "gfOut_trainingfile = '/lu213/brandon.lind/data/testdir/test_gradient_forest_training.RDS'\n",
    "future_range_file = '/data/projects/pool_seq/environemental_data/new_netCDF_files/NA_ENSEMBLE_rcp45_2050s/NA_ENSEMBLE_rcp45_2050s_all-envs_WGS84_clipped_interior.txt'\n",
    "predfile = '/lu213/brandon.lind/data/testdir/test_gradient_forest_predOut.RDS'\n",
    "maskfile = '/data/projects/pool_seq/DF_datasets/DF_pooled_GEA/DF_pooled/snpsANDindels/interior_variety_without-p6/03_maf-p05_RD-recalculated_FDI/gradient_forest/mask.RDS'\n",
    "basename = 'something'\n",
    "save_dir = '/lu213/brandon.lind/data/testdir'\n",
    "\n",
    "!$Rscript $fitting_file $gfOut_trainingfile $future_range_file $predfile $maskfile $basename $save_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
